### PR TITLE
Added geNomad tool for viral classification of assemblies

### PR DESCRIPTION
**Pull request by PhilPalmer**
Hi,

Thanks for developing a great pipeline!

I have been using the nf-core/mag pipeline for a project where I want to investigate what percentage of the assemblies/reads are viral. I noticed that the pipeline is currently more focused on bacteria (eg using the GTD-Tk for taxonomic classification which only supports bacteria/archaea). I, therefore, decided to add a new (optional) process using a recently developed tool [geNomad](https://github.com/apcamargo/genomad) (see [paper](https://academic.oup.com/nar/advance-article/doi/10.1093/nar/gkac1037/6833254)) which performs viral classification i.e. predicts which assemblies are viral and assigns them to taxa.

Would you be interested in including this tool within the pipeline?

Hopefully, the PR should be most of the way there. I have followed the nf-core contributing guidelines, tested the changes myself on a full-scale dataset and added CI tests. Some things that I have not done are:
- [ ] Used a BioContainers Docker/Singularity container (because the tool is not yet available in BioContainers)
- [ ] Updated the workflow schema/image (because I am not sure how you usually do this)

## PR checklist

- [x] This comment contains a description of changes (with reason).
- [x] If you've fixed a bug or added code that should be tested, add tests!
  - [x] If you've added a new tool - have you followed the pipeline conventions in the [contribution docs](https://github.com/nf-core/mag/tree/master/.github/CONTRIBUTING.md)
  ~- [ ] If necessary, also make a PR on the nf-core/mag _branch_ on the [nf-core/test-datasets](https://github.com/nf-core/test-datasets) repository.~
- [x] Make sure your code lints (`nf-core lint`).
- [x] Ensure the test suite passes (`nextflow run . -profile test,docker --outdir <OUTDIR>`).
- [x] Usage Documentation in `docs/usage.md` is updated.
- [x] Output Documentation in `docs/output.md` is updated.
- [x] `CHANGELOG.md` is updated.
- [x] `README.md` is updated (including new tool citations and authors/contributors).

### CODE REVIEW

**Comment by jfy133** [comment]
Ill let Daniel and Sabrina decide on inclusion, but updating the workflow diagram is just done in inkscape. Feel free to ping me if you don't know how/don't have time and I can do it :+1:

**Comment by PhilPalmer** [comment]
Great, thanks @jfy133,

I can try updating the workflow diagram. Should I update the [figure that includes CheckM](https://github.com/nf-core/mag/blob/9cdf461bb6eb861d9ddc8d6be63f49f657e28181/docs/images/mag_workflow.svg)?

Also, I'm not sure why the virus classification CI test fails when running here during the `mmseqs` subprocess. The same command works locally so maybe it's to do with computational resources? I will look into this also

**Comment by jfy133** [comment]
1. Yes, that one!
2. I can't see anything from the error indicating that it's a memory issue..., but I don't have any experience with mmseqs

**Review by d4straub — COMMENTED** [review]
I like the addition!

About the failing tests, probably using a higher value for split (currently 0) should reduce RAM (but I see in the error message no indication that RAM is the issue), according to [here](https://github.com/apcamargo/genomad#executing-genomad):

> The --splits 16 parameter was used here to make it possible to run this example in a notebook.

**Comment by PhilPalmer** [comment]
Hi @d4straub,

Apologies for dropping the ball on this. I haven't worked on this since I submitted the PR and the CI test failed. I will work on fixing the issue and getting the PR to completion this week

**Comment by jfy133** [comment]
Note that it looks like a proper nf-core module is incoming: https://github.com/nf-core/modules/pull/3451

**Comment by CarsonJM** [comment]
Hi @PhilPalmer @jfy133 @d4straub, I just edited the files so that 1) the geNomad nf-core module is used and 2) "virus classification" is changed to "virus identification" in each file because I am planning to add CheckV downstream of geNomad and include it as part of the "virus identification" portion of this pipeline. I just created a PR for @PhilPalmer's naobservatory:genomad branch, does that seem like the best way to incorporate these changes?

**Comment by jfy133** [comment]
PR against `naobservatory` makes sense to me!

**Comment by PhilPalmer** [comment]
Hey @CarsonJM,

Thank you for making these changes! They seem great to me.

I've merged the changes into the `naobservatory:genomad` branch so that they can be reviewed by @jfy133 and @d4straub or others here :)

File: nextflow_schema.json
Comment (d4straub) [inline nextflow_schema.json:497]: Did I miss `params.virus_classification` or is it indeed not used anywhere?

File: nextflow.config
Comment (d4straub) [inline nextflow.config:57]: Those params are not documented in `nextflow_schema.json`?

**Review by d4straub — COMMENTED** [review]
Looks good to me, but documentation seems to need an update?

File: conf/test_virus_identification.config
Comment (prototaxites) [inline conf/test_virus_identification.config:23]: It depends what the purpose of the test is, but would it be valuable to test this with samples that include some viral sequences? AFAIK the mag test samples are simulated from some bacteria (??) so the test may return nothing - worth checking!

Comment (prototaxites) [reply conf/test_virus_identification.config:23]: There's some SARS-COV-2 fastq files in the modules/ test-data repo that might be easily re-purposed here? https://github.com/nf-core/test-datasets/tree/modules/data/genomics/sarscov2/illumina/fastq

File: nextflow.config
Comment (CarsonJM) [reply nextflow.config:57]: I forgot to update the schema after editing @PhilPalmer's code. Thank you for catching this! Schema has been updated and I opened a PR to @PhilPalmer's repo.

File: conf/test_virus_identification.config
Comment (CarsonJM) [reply conf/test_virus_identification.config:23]: Hi @prototaxites thank you for the suggestion! I just looked at the test with the current dataset and it looks like there are some positive hits (most of these are an artifact of the tool's inability to accurately classify very short (<100bp) sequences, but also likely from some prophage artifacts within these bacterial genomes). Do you still think it would be best to change the test data?

Comment (prototaxites) [reply conf/test_virus_identification.config:23]: If it's passing some sequences, I'd say it's probably fine - as long as those can pass through CheckV without error once that's implemented!

File: nextflow_schema.json
Comment (prototaxites) [inline nextflow_schema.json:556]: We were discussing this here (https://github.com/nf-core/mag/issues/455), but for now would it make sense to retain the `skip_{tool}` schematic for consistency?

Comment (CarsonJM) [reply nextflow_schema.json:556]: I think that would make sense! Would you like me to go through and update the code to use `skip_genomad` or would you prefer to get feedback from other maintainers first?

**Comment by CarsonJM** [comment]
Are there any additional changes that I should make before the CI tests are run? (Also, thank you @d4straub and @prototaxites for your previous suggestions!)

File: nextflow.config
Comment (d4straub) [reply nextflow.config:57]: Thanks!

**Comment by d4straub** [comment]
I am happy with the PR.

File: docs/output.md
Comment (jfy133) [inline docs/output.md:298]: Do the names of the files purely have sample/group? Without looking at the rest of the code yet - are you sure these don't need to be output on a per assember+binner+sample basis?

File: modules/nf-core/genomad/download/genomad-download.diff
Comment (jfy133) [inline modules/nf-core/genomad/download/genomad-download.diff:9]: This is out of date, we should update all modules to use the new registries... but I'll makew an issue for this

File: nextflow_schema.json
Comment (jfy133) [reply nextflow_schema.json:632]: I think actually for now we should keep it to opt-in (Even if most of the pipeline is opt out). `gunc` checks are opt-in and the original design for mag was rather for bacteria rather than viruses (although that is expanding).

I think it's safe to leave it as it is and we do a overhaul of the parameters to make them more consistent in 3.0

File: nextflow_schema.json
Comment (jfy133) [inline nextflow_schema.json:573]: In what form? Can it be a folder or must it be `.tar.gz?

File: nextflow_schema.json
Comment (jfy133) [inline nextflow_schema.json:634]: Using what? Will there be other options in teh future?

**Review by jfy133 — APPROVED** [review]
Mostly question about the output docs, but otherwise LGTM!

File: README.md
Comment (jfy133) [reply README.md:37]: And the pipeline diagram needs to be updateD!

File: workflows/mag.nf
Comment (prototaxites) [inline workflows/mag.nf:631]: One other minor point - would it make sense to give this subworkflow a sematic name (e.g. VIRUS_IDENTIFICATION) in case we want to add other tools down the line that aren't Genomad?

Comment (jfy133) [reply workflows/mag.nf:631]: Good point!

File: docs/output.md
Comment (CarsonJM) [reply docs/output.md:298]: geNomad's output files are named based on the input file (i.e. input = test1.fasta; outputs = test1_virus_summary.tsv). If you think it is necessary to name outputs based on assembler+binner+sample, I could make a patch to the geNomad module where the output files are all renamed!

File: README.md
Comment (CarsonJM) [inline README.md:36]: So are you saying that line 37 should be appended to line 36?
```suggestion
- assigns taxonomy to bins using [GTDB-Tk](https://github.com/Ecogenomics/GTDBTk) and/or [CAT](https://github.com/dutilh/CAT) and optionally identifies viruses in assemblies using [geNomad](https://github.com/apcamargo/genomad)
```

File: nextflow_schema.json
Comment (CarsonJM) [inline nextflow_schema.json:573]: It should actually be a folder! So it should look something like this?
```suggestion
                "genomad_db": {
                    "type": "string",
                    "default": "https://portal.nersc.gov/genomad/__data__/genomad_db_v1.1.tar.gz",
                    "description": "Database for virus classification with geNomad",
                    "help_text": "Must be a directory containing the uncompressed contents from https://portal.nersc.gov/genomad/__data__/genomad_db_v1.1.tar.gz"
```

File: nextflow_schema.json
Comment (CarsonJM) [reply nextflow_schema.json:634]: No other options for virus identification planned at the moment. Only perhaps CheckV (which checks virus quality) which will be downstream.
```suggestion
                    "description": "Identify viral sequences in assemblies using geNomad"
```

File: README.md
Comment (CarsonJM) [reply README.md:37]: Updated pipeline diagram pushed to @PhilPalmer's branch!

File: workflows/mag.nf
Comment (CarsonJM) [reply workflows/mag.nf:631]: Workflow has been renamed in PR I submitted to @PhilPalmer's fork!

File: README.md
Comment (jfy133) [reply README.md:36]: I think so yes :)

File: nextflow_schema.json
Comment (jfy133) [reply nextflow_schema.json:634]: Yes, fine for now :+!

File: docs/output.md
Comment (prototaxites) [reply docs/output.md:298]: Poking about - modules.conf doesn't specify `ext.prefix`, so the output files should just be named `[sample/group]_*.*` - might run into overwriting of results if more than one assembler runs. Should fix by specifying `ext.prefix = { "${meta.id}-${meta.assembler}" }` or similar in the config file. This runs before binning so there should be no need to specify the binner too.

File: docs/images/mag_workflow.png
Comment (prototaxites) [inline docs/images/mag_workflow.png:1]: I think because the PR was started a long time ago, the version of the workflow diagram you modified is a bit out of date - it's missing domain classification. Can you modify the new version instead?

https://github.com/nf-core/mag/blob/dev/docs/images/mag_workflow.svg

File: workflows/mag.nf
Comment (CarsonJM) [inline workflows/mag.nf:632]:
```suggestion
Sorry, just caught this error!
        ch_versions = ch_versions.mix(VIRUS_IDENTIFICATION.out.versions.first())
```

File: workflows/mag.nf
Comment (CarsonJM) [inline workflows/mag.nf:632]:
```suggestion
```

File: docs/images/mag_workflow.png
Comment (CarsonJM) [reply docs/images/mag_workflow.png:1]: nf-core-mag genomad_schematic #2

@prototaxites Does this version look correct to you?

Comment (prototaxites) [reply docs/images/mag_workflow.png:1]: LGTM!

File: docs/output.md
Comment (jfy133) [reply docs/output.md:298]: @prototaxites I think though was the purpose of this comment

> geNomad's output files are named based on the input file

So there is no way to specify the output explicitly? Therefore I _assume_ that as the input file _should_ have the assembly/binner name, then we don't need to use `prefix` and manual `mv` commands in the module. However this is somethign for @CarsonJM to confirm

File: README.md
Comment (jfy133) [inline README.md:37]:
```suggestion
```

**Review by jfy133 — APPROVED** [review]
Just need the image updated, then it's na OK from me

File: docs/output.md
Comment (prototaxites) [reply docs/output.md:298]: Ah, good point - I was thrown because `prefix` is defined in the module, but it turns out it's not actually used within the script block. In that case, I agree we shouldn't see any conflicts - but perhaps worth updating the docs to reflect the output file names?

Comment (jfy133) [reply docs/output.md:298]: Agreed! @CarsonJM please double check and update as necessary :pray:

**Comment by jfy133** [comment]
@prototaxites @d4straub if we get a :heavy_check_mark:  from you, then I guess we are good to go for a merge?

File: workflows/mag.nf
Comment (prototaxites) [inline workflows/mag.nf:641]: One last little one - should we now call this `params.run_virus_identification`?

**Review by prototaxites — APPROVED** [review]
✅ from me in general! One last little comment, and I think the output docs still need updating to make sure they list the correct info.

File: workflows/mag.nf
Comment (CarsonJM) [reply workflows/mag.nf:641]: Yep, that makes sense to me. Sorry for missing that on the last pass!

File: docs/output.md
Comment (CarsonJM) [reply docs/output.md:298]: The current output looks like this:
`MEGAHIT-group-0.contigs_summary`

So should the output docs say something like:
`VirusIdentification/geNomad/[assembler]-[sample/group]_summary`

The only part I'm unsure about is that the `.contigs` would be `_scaffolds` if metaSPAdes is chosen as the assembler

Comment (prototaxites) [reply docs/output.md:298]: I think you could just throw a wildcard `[assembler]-[sample/group]_*_summary` in there to catch the difference between `.contigs` and `_scaffolds`. Not a perfect solution but I think it's clear enough!

**Review by d4straub — APPROVED** [review]
LGTM

**Comment by jfy133** [comment]
@CarsonJM once the PR to the fork is in, and tests pass here, you're welcome to merge :)

**Comment by CarsonJM** [comment]
🚀🚀🚀
Thanks for all the help with this PR!

**Pull request merged**
